### PR TITLE
Support referencing `Secret`s and `ConfigMap`s in `Extension` and `ControllerDeployment` Helm values

### LINKED ISSUE #14682: Support Resource References in `Extension` and `ControllerDeployment` Helm Values
**How to categorize this issue?**

/area control-plane
/area security
/kind enhancement
/kind api-change

**What would you like to be added**:

Add support for referencing `Secret`s and `ConfigMap`s in the Helm chart `values` fields of the following API objects, similar to how `Shoot`, `Seed`, and `Garden` resources already support referenced resources via the `resources` field:

1. **`core.gardener.cloud/v1.ControllerDeployment`**:
   - Add a `resources` field to reference Secrets/ConfigMaps
   - Allow referencing these resources in `spec.helm.values` using template substitution (Option 1) or JSONPath references (Option 2)

2. **`operator.gardener.cloud/v1alpha1.Extension`**:
   - Add a `resources` field to reference Secrets/ConfigMaps
   - Allow referencing these resources in:
     - `spec.deployment.extension.values`
     - `spec.deployment.extension.runtimeClusterValues`
     - `spec.deployment.admission.values`

**Proposed API changes:**

Add a `resources []NamedResourceReference` field to both API types (following the existing pattern from `Shoot`/`Seed`/`Garden` APIs).

**Resolution mechanism - Two possible approaches:**

The resolution of resource references in `values` should be handled by the respective controllers:
- For `ControllerDeployment`: `gardenlet` resolves references when applying charts to seed clusters
- For `Extension`: `gardener-operator` resolves references when applying charts to the runtime cluster

### Option 1: Template Substitution (Go templates)

Use Go template syntax in the values to reference resource data.

**ControllerDeployment using template syntax:**
```yaml
apiVersion: core.gardener.cloud/v1
kind: ControllerDeployment
metadata:
  name: my-extension
spec:
  helm:
    ociRepository:
      repository: europe-docker.pkg.dev/gardener-project/releases/extensions/my-extension
      tag: v1.0.0
    values:
      credentials:
        username: "{{ .resources.my-credentials.data.username }}"
        password: "{{ .resources.my-credentials.data.password }}"
      config: "{{ .resources.my-config.data.config\\.yaml }}"
      featureFlags: "{{ .resources.my-config.data.feature-flags }}"
  resources:
  - name: my-credentials
    resourceRef:
      apiVersion: v1
      kind: Secret
      name: extension-credentials
  - name: my-config
    resourceRef:
      apiVersion: v1
      kind: ConfigMap
      name: extension-config
```

### Option 2: JSONPath References

Use JSONPath syntax in the reference resource config.

**ControllerDeployment using JSONPath syntax:**
```yaml
apiVersion: core.gardener.cloud/v1
kind: ControllerDeployment
metadata:
  name: my-extension
spec:
  helm:
    ociRepository:
      repository: europe-docker.pkg.dev/gardener-project/releases/extensions/my-extension
      tag: v1.0.0
    values: {}
  resources:
  - name: my-credentials
    resourceRef:
      apiVersion: v1
      kind: Secret
      name: extension-credentials
    valuesPath:
    - dataKey: username
       valuesPath: .credentials.username
    - dataKey: password
       valuesPath: .credentials.password
  - name: my-config
    resourceRef:
      apiVersion: v1
      kind: ConfigMap
      name: extension-config
    valuesPath:
    - dataKey: config.yaml
       valuesPath: .config
    - dataKey: feature-flags
       valuesPath: .featureFlags
```

**Why is this needed**:

1. **Security**: Storing sensitive data (credentials, tokens, certificates) directly in the `values` field is not ideal. Secret references allow proper secret management with:
   - Encryption at rest
   - RBAC controls
   - Audit logging
   - Secret rotation capabilities

2. **External Management**: Organizations often manage secrets through external systems (e.g., external secret operators, credential rotation controllers, vault). This is especially not supported if `Extension` or `ControllerDeployment` resources are managed via GitOps approaches.

3. **Credential Rotation**: Automated credential rotation is difficult when credentials are inlined. With Secret references:
   - A rotation controller updates only the Secret
   - The extension controller picks up new credentials on its next reconciliation
   - No modification to the `ControllerDeployment` or `Extension` object is required (it can be checked-in as such into a Git repository)
   - The rotation is transparent to the deployment specification

**Backwards compatibility**:

- The existing `values` fields remain unchanged and continue to work as-is
- The new `resources` field is optional
- Existing deployments continue to work without modification
- When `resources` are specified, the controller resolves references before applying the chart
- No breaking changes to the API

### PR DESCRIPTION
**Pull request by oliver-goetz**
**How to categorize this PR?**

/area control-plane
/kind enhancement 
                                                                                                                                                                                 
**What this PR does / why we need it**:

This PR adds support for referencing `Secret`s and `ConfigMap`s from Helm values of `Extension` (`operator.gardener.cloud`) and `ControllerDeployment` (`core.gardener.cloud`) resources. This allows extension operators to keep credentials and configuration in dedicated `Secret`s/`ConfigMap`s and template them into the Helm values via Go template expressions like `{{ .resources.<name>.data.<key> }}`, instead of inlining sensitive or environment-specific data into
 the resource itself.

Details:
- New API fields:
  - `Extension.spec.deployment.resources` (resolved by `gardener-operator` against the garden runtime cluster). Usable in `spec.deployment.extension.values`, `spec.deployment.extension.runtimeClusterValues`, and `spec.deployment.admission.values`.
  - `ControllerDeployment.resources` (resolved by `gardenlet` against the virtual garden cluster). Usable in `helm.values`.
  - `ControllerInstallation.spec.resourceRefs` is written by `gardener-controller-manager` based on `ControllerDeployment.resources` and consumed by `gardenlet`.
- Only `Secret` and `ConfigMap` references with `apiVersion: v1` are supported. Referenced objects must reside in the `garden` namespace and carry the label `gardener.cloud/role: resource-reference`.
- Admission validation enforces the strict template form `{{ .resources.<name>.data.<key> }}` and rejects references to undeclared `<name>`s.
- The `controllerinstallation` controllers in `gardenlet` and the `extension` controller in `gardener-operator` watch the referenced resources and trigger reconciliation on changes.
- The seed authorizer graph was extended so that `gardenlet` is authorized to read the referenced `Secret`s/`ConfigMap`s.
- `provider-local` was adjusted to use the new mechanism as a usage example.
- `docs/extensions/registration.md` was extended with a new "Referencing Resources in Helm Values" section.

**Which issue(s) this PR fixes**:
Fixes #14682

**Special notes for your reviewer**:
/cc @MartinWeindel @timuthy 

**Release note**:
```other operator
`Extension` (`operator.gardener.cloud`) and `ControllerDeployment` (`core.gardener.cloud`) resources can now reference `Secret`s and `ConfigMap`s from the `garden` namespace via the new `resources` field. The data of referenced objects can be templated into Helm values using `{{ .resources.<name>.data.<key> }}`. Referenced objects must carry the label `gardener.cloud/role: resource-reference`.
```

### CODE REVIEW

File: pkg/utils/chart/values.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gardencorev1 "github.com/gardener/gardener/pkg/apis/core/v1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/utils"
+)
+
+// resourceNameRegex searches for resources referenced via `{{ .resources.<name>.data.<key> }}`.
+var resourceNameRegex = regexp.MustCompile(`\{\{-?\s*\.resources\.(?P<name>[^.\s}]+)\.data\.[^.\s}]+\s*-?\}\}`)
```
Comment (MartinWeindel) [inline pkg/utils/chart/values.go]: The regex is more permissive than the regex in the validation (file pkg/api/core/validation/utils.go, line 699). Consider to use only one regex and use them at both places.

File: pkg/utils/chart/values.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, dataMap); err != nil {
+		return "", fmt.Errorf("failed to execute template %q: %w", s, err)
+	}
+	return buf.String(), nil
+}
+
+// ResourceNamesFromValues returns the names of all resources referenced in the given values via template
+// expressions like {{ .resources.<name>.data.<key> }}.
+func ResourceNamesFromValues(values *apiextensionsv1.JSON) sets.Set[string] {
```
Comment (MartinWeindel) [inline pkg/utils/chart/values.go]: From a pragmatic view, the method does its job. But edge cases return more names than expected.
Example:

```
{"{{ .resources.foo.data.bar}}": 42, "first": "{{ .resources.first.data.","second":"second}}"}
```
returns `{"foo","first"}`

I suggest to generalize the substituteAny method to a visitor pattern and use it for substitution and name extraction.

File: pkg/utils/chart/values_test.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+		It("should extract resource names from nested values", func() {
+			values := &apiextensionsv1.JSON{Raw: []byte(`{"logLevel":"{{ .resources.logLevels.data.extension }}","imageVectorOverwrite":{"images":[{"name":"ccm","ref":"{{ .resources.imageRefs.data.ccm }}"},{"name":"mcm","ref":"{{ .resources.imageRefs.data.mcm }}"}]}}`)}
+			Expect(ResourceNamesFromValues(values)).To(Equal(sets.New("logLevels", "imageRefs")))
+		})
+
+		It("should deduplicate resource names", func() {
+			values := &apiextensionsv1.JSON{Raw: []byte(`{"a":"{{ .resources.foo.data.x }}","b":"{{ .resources.foo.data.y }}"}`)}
+			Expect(ResourceNamesFromValues(values)).To(Equal(sets.New("foo")))
+		})
+	})
```
Comment (MartinWeindel) [inline pkg/utils/chart/values_test.go:251]: Add test case for malformed templates:
```
		It("should not extract resource names from malformed templates", func() {
			values := &apiextensionsv1.JSON{Raw: []byte(`"{{ .resources.foo.data.bar}}": 42,"first": "{{ .resources.first.data.","second":"second}}"`)}
			// should NOT return `foo` or ´first` or `second` because the templates are malformed and thus not valid references
			Expect(ResourceNamesFromValues(values)).To(Equal(sets.New[string]()))
		})
```

**Review by MartinWeindel — CHANGES_REQUESTED** [review]
Thanks for implementing support for secret and configmap resource references!
Implementation, documentation and test coverage look good to me.
I have some minor concerns about regex usage and the `ResourceNamesFromValues`.
In the next days, I want to make a second round of review by using it in the shoot-cert-service.

File: pkg/utils/chart/values.go (hunk 1 above)
Comment (oliver-goetz) [reply pkg/utils/chart/values.go]: Done, the strict version is used for both cases.

File: pkg/utils/chart/values.go (hunk 2 above)
Comment (oliver-goetz) [reply pkg/utils/chart/values.go]: Okay, done 🙂 
While doing this I also allowed template for keys of a map. This way it is possible to template typical `labels` and `annotations` sections of helm values too.

File: pkg/utils/chart/values_test.go (hunk 3 above)
Comment (oliver-goetz) [reply pkg/utils/chart/values_test.go:251]: Done. I also found some uncovered edge cases in the api validation and added tests for them too.

File: pkg/apis/operator/v1alpha1/types_extension.go [hunk 4]
```diff
@@ -65,6 +65,15 @@ type Deployment struct {
...
 	// +optional
 	AdmissionDeployment *AdmissionDeploymentSpec `json:"admission,omitempty"`
+	// Resources is a list of named resource references that can be referenced in the Helm chart values via Go
+	// template syntax (e.g. `{{ .resources.<name>.data.<key> }}`). Only resources of kind `Secret` and `ConfigMap`
+	// (apiVersion `v1`) are supported. The referenced resources must reside in the garden namespace.
+	// References can be used in `spec.deployment.extension.values`, `spec.deployment.extension.runtimeClusterValues`
+	// and `spec.deployment.admission.values`.
+	// +optional
+	// +patchMergeKey=name
+	// +patchStrategy=merge
```
Comment (timuthy) [inline pkg/apis/operator/v1alpha1/types_extension.go]: Side note: We probably don't need this here since strategic merge patches are not supported for CRDs.

File: pkg/utils/graph/eventhandler_controllerdeployment.go
Comment (timuthy) [inline pkg/utils/graph/eventhandler_controllerdeployment.go:1]: Could you also update https://github.com/gardener/gardener/blob/master/docs/deployment/gardenlet_api_access.md?

**Review by timuthy — COMMENTED** [review]
Thank you very much for the enhancement, I really like it!

The changes look good after having a first look. Besides the comments, could you please enhance an E2E test to cover the secret and configmap reference?

File: pkg/api/operator/v1alpha1/validation/extension.go [hunk 5]
```diff
@@ -42,25 +59,33 @@ func validateDeployment(deployment *operatorv1alpha1.Deployment, fldPath *field.
 		return append(allErrs, field.Required(fldPath, "at least one of extension or admission must be specified"))
 	}
 
-	allErrs = append(allErrs, validateExtensionDeployment(deployment.ExtensionDeployment, fldPath.Child("extension"))...)
-	allErrs = append(allErrs, validateAdmissionDeployment(deployment.AdmissionDeployment, fldPath.Child("admission"))...)
+	coreResources, resourceErrs := validateResourceReferences(deployment.Resources, fldPath.Child("resources"))
```
Comment (RadaBDimitrova) [inline pkg/api/operator/v1alpha1/validation/extension.go]: To me this looks like we're doing 2 things in the `validateResourceReferences` func - validating and converting to get the resources. I think it would make more sense to split the two to keep the functions simple and follow the convention. You could also not split it, but maybe rename the function to signify the dual responsibilities of the function.

File: pkg/api/operator/v1alpha1/validation/extension.go [hunk 6]
```diff
@@ -42,25 +59,33 @@ func validateDeployment(deployment *operatorv1alpha1.Deployment, fldPath *field.
...
+	allErrs = append(allErrs, validateExtensionDeployment(deployment.ExtensionDeployment, coreResources, fldPath.Child("extension"))...)
+	allErrs = append(allErrs, validateAdmissionDeployment(deployment.AdmissionDeployment, coreResources, fldPath.Child("admission"))...)
+	allErrs = append(allErrs, resourceErrs...)
```
Comment (RadaBDimitrova) [inline pkg/api/operator/v1alpha1/validation/extension.go]: This should be right after `validateResourceReferences`.

**Review by RadaBDimitrova — CHANGES_REQUESTED** [review]
Initial review, will continue later

File: pkg/api/core/validation/utils.go [hunk 7]
```diff
@@ -688,6 +688,58 @@ func ValidateResources(resources []core.NamedResourceReference, fldPath *field.P
...
+	// resourceReferenceRegexp validates the only allowed template form: `{{ .resources.<name>.data.<key> }}`
+	// where <name> and <key> are alphanumeric and surrounding whitespace is optional. The first capture group
+	// extracts <name> so callers can verify it against the declared resources list.
+	resourceReferenceRegexp = regexp.MustCompile(`^\{\{\s*\.resources\.([a-zA-Z0-9]+)\.data\.[a-zA-Z0-9]+\s*\}\}$`)
+)
+
+// ValidateAlphanumericResourceNames ensures the resource names are alphanumeric. An empty name also fails this check
+// (in addition to the Required error from ValidateResources) because there is no way to create a reference to an
+// empty name.
+func ValidateAlphanumericResourceNames(resources []core.NamedResourceReference, fldPath *field.Path) field.ErrorList {
```
Comment (oliver-goetz) [inline pkg/api/core/validation/utils.go:703]: Go templates support alphanumeric variable names only. I added this as a comment.

File: pkg/operator/controller/extension/extension/runtime/runtime.go [hunk 8]
```diff
@@ -107,6 +108,17 @@ func (d *deployer) createOrUpdateResources(ctx context.Context, extension *opera
...
+	if len(extension.Spec.Deployment.Resources) > 0 && helmValues != nil {
+		resources, err := chartutils.ResolveResources(ctx, d.runtimeClientSet.Client(), d.gardenNamespace, extension.Spec.Deployment.Resources)
+		if err != nil {
+			return fmt.Errorf("failed resolving resource references: %w", err)
+		}
+		helmValues, err = chartutils.SubstituteTemplateInValues(helmValues, resources)
+		if err != nil {
+			return fmt.Errorf("failed substituting resource references in extension runtimeClusterValues: %w", err)
+		}
+	}
```
Comment (RadaBDimitrova) [inline pkg/operator/controller/extension/extension/runtime/runtime.go]: Maybe extract this in a helper function, since the logic is used in a bunch of places, so as to avoid code duplication.

File: pkg/utils/chart/values.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	return result, nil
+}
+
+// SubstituteTemplateInValues walks the given values structure and renders any string (both map keys and
+// values) as Go templates with the provided resource references as template data. Non-string scalars are
+// left untouched. The resulting structure has the same shape as the input. Templates can reference
+// resolved data via `{{ .resources.<name>.data.<key> }}`.
+func SubstituteTemplateInValues(values map[string]any, resources Resources) (map[string]any, error) {
```
Comment (maboehm) [inline pkg/utils/chart/values.go:101]: this approach conflicts with helm charts that use `tpl` to render dynamic values:
https://helm.sh/docs/howto/charts_tips_and_tricks/#using-the-tpl-function
So it might be totally valid to have `foo: "{{.Release.Name }}"`. With your current approach, authors would need to write ``foo: "{{ `{{ .Release.Name }}` }}"``

I think this is probably an acceptable limitation (I doubt any gardener extension charts use it), but this should be documented. (especially since the behavior changes when going from 0 referenced resources to 1). 

If we think this is a problem, we could adapt a similar API design as flux, that uses a
```yaml
valuesFrom:
  - kind: ConfigMap
    name: generic-token-kubeconfig-name
    targetPath: extraVolumes[0].projected.sources[0].secret.name
    valuesKey: GENERIC_TOKEN_KUBECONFIG_SECRET_NAME
```
and this helper to render it: https://github.com/fluxcd/pkg/blob/main/chartutil/values.go#L261

Comment (oliver-goetz) [reply pkg/utils/chart/values.go:101]: This change allows templating `values.yaml` only. Afaik in Helm this file is a static yaml file where you can't use any templating function.
The Go templating introduced in this PR is executed before the values are passed to the Helm, so there shouldn't be an issue with the `tpl` function, should it?

Btw. we voted in #14682 in favor of the templating approach 😅

Comment (MartinWeindel) [reply pkg/utils/chart/values.go:101]: Another options could be to use different delimiters for the Go templating, e.g. `<<` and `>>`

```go
func renderTemplate(s string, data templateData) (string, error) {
	tmpl, err := template.New("values").Delims("<<", ">>").Option("missingkey=error").Parse(s)
```
A similar approach is used in the [shoot-falco-service extension](https://github.com/gardener/gardener-extension-shoot-falco-service/blob/bdcb8e740cfc15e9c0dff3a1fda5dc2ac99c75a0/pkg/values/falcovalues.go#L1174)

File: pkg/operator/controller/extension/reference/add.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+	var names []string
+	for _, ref := range extension.Spec.Deployment.Resources {
+		if ref.ResourceRef.APIVersion == apiVersion && ref.ResourceRef.Kind == kind {
+			names = append(names, ref.ResourceRef.Name)
+		}
+	}
+	return names
+}
+
+func namesForPullRequestSecrets(extension *operatorv1alpha1.Extension) []string {
```
Comment (RadaBDimitrova) [inline pkg/operator/controller/extension/reference/add.go]: Is this not supposed to be pull secrets for OCI images, maybe this is mistyped?

**Review by RadaBDimitrova — COMMENTED** [review]
Overall nice PR, just some nits

**Review by RadaBDimitrova — COMMENTED** [review]
Thanks for the quick adaptations
/lgtm

File: dev-setup/extensions/provider-local/components/extension-with-references/patch-extension-prow.yaml
Comment (timuthy) [inline dev-setup/extensions/provider-local/components/extension-with-references/patch-extension-prow.yaml:1]: Where is this patch used?

File: docs/deployment/gardenlet_api_access.md [hunk 11]
```diff
@@ -50,7 +50,7 @@ Today, the following rules are implemented:
...
-| `ConfigMap`                 | `get`                                                           | `ConfigMap` -> `Seed`, `ConfigMap` -> `Shoot` -> `Seed`                                                                                                                              | Allow only `get` requests for `ConfigMap`s referenced by `Seed`s, or `Shoot`s that are assigned to the `gardenlet`'s `Seed`. Allow reading the `kube-system/cluster-identity` `ConfigMap`.                                                                                      |
+| `ConfigMap`                 | `get`                                                           | `ConfigMap` -> `Seed`, `ConfigMap` -> `Shoot` -> `Seed`, `ConfigMap` -> `ControllerDeployment` -> `ControllerInstallation` -> `Seed`                                                 | Allow only `get` requests for `ConfigMap`s referenced by `Seed`s, `Shoot`s assigned to the `gardenlet`'s `Seed`, or by `ControllerDeployment`s of `ControllerInstallation`s assigned to the `gardenlet`'s `Seed`. Allow reading the `kube-system/cluster-identity` `ConfigMap`. |
```
Comment (timuthy) [inline docs/deployment/gardenlet_api_access.md:53]: Can you adapt `docs/deployment/content/gardenlet_api_access_graph.png` as well?

File: pkg/utils/gardener/garden.go
Comment (timuthy) [inline pkg/utils/gardener/garden.go:1]: This is a fix independent from the feature being added in this PR, right? Should we cherry-pick it afterwards?

File: pkg/apis/config/controllermanager/v1alpha1/types.go [hunk 12]
```diff
@@ -58,6 +58,9 @@ type ControllerManagerControllerConfiguration struct {
 	// ControllerDeployment defines the configuration of the ControllerDeployment controller.
 	// +optional
 	ControllerDeployment *ControllerDeploymentControllerConfiguration `json:"controllerDeployment,omitempty"`
+	// ControllerDeploymentReference defines the configuration of the ControllerDeploymentReference controller. If unspecified, it is defaulted with `concurrentSyncs=5`.
```
Comment (timuthy) [inline pkg/apis/config/controllermanager/v1alpha1/types.go]: > If unspecified, it is defaulted with `concurrentSyncs=5`

Should probably be added to the `ConcurrentSyncs` field in the struct. I think you can even drop it entirely or add the same comment to the other concurrent sync fields for the sake of consistency.

File: docs/extensions/registration.md [hunk 13]
```diff
@@ -311,6 +311,57 @@ In order to allow extension controller deployments to get information about the
...
+
+The following references are supported:
+
+- `Extension.spec.deployment.resources` (resolved by `gardener-operator` against the garden runtime cluster). The references can be substituted in `Extension.spec.deployment.extension.values`, `Extension.spec.deployment.extension.runtimeClusterValues`, and `Extension.spec.deployment.admission.values`.
+- `ControllerDeployment.resources` (resolved by `gardenlet` against the virtual garden cluster). The references can be substituted in `ControllerDeployment.helm.values`.
+
+Only `Secret` and `ConfigMap` references with `apiVersion: v1` are supported.
+For `Secret`s the values of `data` are base64-decoded; for `ConfigMap`s the `data` entries are exposed under `data`.
+
+The referenced object must additionally carry the label `gardener.cloud/role: resource-reference` so that the controllers watching them can pick up changes and trigger reconciliations.
```
Comment (timuthy) [inline docs/extensions/registration.md:327]: Didn't you want to remove this label requirement, including `helm-pull-secret` and `oci-ca-bundle`?

File: docs/concepts/operator.md
Comment (timuthy) [inline docs/concepts/operator.md:1]: Niice!

File: pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go [hunk 14]
```diff
@@ -199,19 +199,10 @@ func (r *Reconciler) reconcile(
...
-		if helmValues != nil {
-			helmValues, err = chartutils.SubstituteTemplateInValues(helmValues, resources)
-			if err != nil {
-				conditionValid = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditionValid, gardencorev1beta1.ConditionFalse, "ChartInformationInvalid", fmt.Sprintf("chart values cannot be templated: %+v", err))
-				return reconcile.Result{}, err
-			}
-		}
+	helmValues, err := chartutils.RenderHelmValues(gardenCtx, r.GardenClient, helmValues, v1beta1constants.GardenNamespace, controllerDeployment.Resources)
+	if err != nil {
+		conditionValid = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditionValid, gardencorev1beta1.ConditionFalse, "RenderingHelmValuesFailed", fmt.Sprintf("helm values cannot be rendered: %+v", err))
```
Comment (timuthy) [inline pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go:204]: Can we cover this case in the integration test?

**Review by timuthy — COMMENTED** [review]
Thanks a lot for the feature and changes 👏

File: docs/deployment/gardenlet_api_access.md (hunk 11 above)
Comment (oliver-goetz) [reply docs/deployment/gardenlet_api_access.md:53]: Do you know how the image has been edited (you changed it for the last time 😉 )? I tried to open it with ExcaliDraw but it did not work.

File: pkg/utils/gardener/garden.go
Comment (oliver-goetz) [reply pkg/utils/gardener/garden.go:1]: Yes, we could cherry-pick the commit. I did not check if it has a negative impact without the changes introduced in this PR though.

File: docs/extensions/registration.md (hunk 13 above)
Comment (oliver-goetz) [reply docs/extensions/registration.md:327]: Finally, I removed the exception for the `gardener.cloud/role` label from reference controller (https://github.com/gardener/gardener/pull/14979/changes/f809be03f861041a6c0d1d9c7c2cdfd510029143) and kept the labels for all the use cases.
However, I missed one part which prevents those secrets from being copied into the seed namespaces. This is fixed now.

File: docs/deployment/gardenlet_api_access.md (hunk 11 above)
Comment (timuthy) [reply docs/deployment/gardenlet_api_access.md:53]: I must have forgotten to include the scene and recreated it 🙂 You can download it from below:

gardenlet_api_access_graph

Comment (oliver-goetz) [reply docs/deployment/gardenlet_api_access.md:53]: Done, I updated my last commit.

File: docs/extensions/registration.md (hunk 13 above)
Comment (timuthy) [reply docs/extensions/registration.md:327]: Thank you! I thought we want to remove the label requirement entirely, so that operators can just refer to a `ConfigMap` or `Secret` via the `Extension`/`ControllerDeployment`, similar to how users do it for shoots.

Let's discuss this offline.

Comment (timuthy) [reply docs/extensions/registration.md:327]: As discussed, we will keep the label requirement for now and create an issue to follow up on making the resource retrieval from the `garden` namespace consistent.

**Review by timuthy — COMMENTED** [review]
/lgtm
/approve
